rework test to use libhostspriv.so

bobvanderlinden opened the pull request:
Currently `hosts.c` includes tests using the `COMPILE_FOR_TEST` define.
It only tests for `lookup_alias`, but not the `LD_PRELOAD` hooking.

This PR attempts to move the test code to a separate executable, which is used with `LD_PRELOAD` and `test_hosts` to test the full flow of libhostspriv.

Currently it only tests for `getaddrinfo`.

figiel commented on test_getaddrinfo.c:
```diff
@@ -0,0 +1,58 @@
+
+	int s = getaddrinfo(hostname, NULL, &hints, &result);
+	if (s != 0) {
+		out[0] = '\0';
```
maybe just initialize result in assert_resolve.

bobvanderlinden replied:
Done :+1:

figiel commented on hosts.c:
```diff
@@ -134,84 +122,4 @@ FUN_BEGIN(gethostbyname2_r, int, name, char *name, int a, void *b, void *c, size
-}
-
-#endif /* COMPILE_FOR_TEST */
+FUN_END(name, a);
```
nit: please keep the newline at the end of file.

bobvanderlinden replied:
Fixed this and added .editorconfig to avoid similar future problems.

figiel commented on test_getaddrinfo.c:
```diff
@@ -0,0 +1,58 @@
+	}
+
+	struct sockaddr_in *addr;
+	addr = (struct sockaddr_in *)result->ai_addr;
```
Why assume IPv4 here?

bobvanderlinden replied:
The example I looked up didn't use `inet_ntop`. I now migrated to use `inet_ntop`, which does support ipv6 :+1: Added ipv6 tests as well.

figiel commented on test_getaddrinfo.c:
```diff
@@ -0,0 +1,58 @@
+	struct addrinfo hints;
+	struct addrinfo *result;
+
+	memset(&hints, 0, sizeof(hints));
```
I don't think this is needed just initialize hints in declaration (e.g. hints = {}).

bobvanderlinden replied:
Used `{}` to initialize the fields as well, much cleaner indeed.

figiel commented on test_getaddrinfo.c:
```diff
@@ -0,0 +1,58 @@
+	hints.ai_flags = 0;
+	hints.ai_protocol = 0;          /* Any protocol */
+
+	int s = getaddrinfo(hostname, NULL, &hints, &result);
```
`s` is not used anywhere else and doesn't mean much, maybe just inline getaddrinfo in if.

bobvanderlinden replied:
Indeed, inlined it now.

bobvanderlinden commented:
I also added a test that 'redirects' hostname resolving `localhost localhostalias`, as I added that in the readme of #6 as well.

bobvanderlinden commented:
Incorporated your comments. I think it looks much better.

figiel commented on test_getaddrinfo.c:
```diff
@@ -0,0 +1,78 @@
+ assert(strcmp(str1, str2)==0); } while (0)
+#define ASSERT assert
+
+const void *getaddr(const struct addrinfo *info) {
```
This likely needs to be static, sorry missed this earlier.

figiel commented on test_getaddrinfo.c:
```diff
@@ -0,0 +1,78 @@
+  if (info->ai_family == AF_INET) {
+    const struct sockaddr_in *addr = (struct sockaddr_in *)info->ai_addr;
+    return &addr->sin_addr;
+  } else if (info->ai_family == AF_INET6) {
```
I'd remove else here, previous if always returns.

figiel commented:
This looks good, I made some nitpick comments, please fix if you agree and squash the commits.

bobvanderlinden commented:
Thanks for the comments! I wasn't sure why the getaddr needed to be static, but i added it anyway :+1:

I reworked the commits a bit to have a cleaner history.

figiel commented:
Static is preferred by default unless you need to access the symbol from another file, otherwise you risk name conflicts.
Thanks for contributions!

bobvanderlinden commented:
Thank you for the explanation and the merge! :smile::+1:
